platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+				   struct vfio_region_info_cap_sparse_mmap
+				   **sparse)
+{
+	*sparse =
+	    odp_container_of(hdr, struct vfio_region_info_cap_sparse_mmap,
+			     header);
+}
+
+static struct vfio_info_cap_header *
+vfio_get_region_info_cap(struct vfio_region_info *info, __u16 id)
+{
+	struct vfio_info_cap_header *hdr;
+	void *ptr = info;
+
+	if (!(info->flags & VFIO_REGION_INFO_FLAG_CAPS))
+		return NULL;
+
+	for (hdr =
+	     (struct vfio_info_cap_header *)((char *)ptr + info->cap_offset);
+	     hdr != ptr;
+	     hdr = (struct vfio_info_cap_header *)((char *)ptr + hdr->next)) {
+		if (hdr->id == id)
+			return hdr;
+	}
+
+	return NULL;
+}
+
+static struct vfio_info_cap_header *
+vfio_get_cap_info(struct vfio_region_info *region_info, __u16 id)
+{
+	struct vfio_info_cap_header *caps = NULL;
+
+	caps = vfio_get_region_info_cap(region_info, id);
+
+	return caps;
+}
+
+int vfio_get_region_sparse_mmaps(struct vfio_region_info *region_info,
+				 struct vfio_region_info_cap_sparse_mmap
+				 **sparse)
+{
+	struct vfio_info_cap_header *caps = NULL;
+	int ret = -ENOENT;
+
+	if (region_info->flags & VFIO_REGION_INFO_FLAG_CAPS &&
+	    region_info->argsz > sizeof(*region_info)) {
+		caps = vfio_get_cap_info(region_info,
+					 VFIO_REGION_INFO_CAP_SPARSE_MMAP);
+		if (!caps)
+			goto out;
+		vfio_find_sparse_mmaps(caps, sparse);
+		if (*sparse) {
+			for (uint32_t i = 0; i < (*sparse)->nr_areas; i++)
+				ODP_DBG("Sparse region: %d 0x%llx %llu\n", i,
+					(*sparse)->areas[i].offset,
+					(*sparse)->areas[i].size);
+		}
+
+		ret = 0;
+	}
+
+out:
+	return ret;
+}
+
+/** Match capability type
+ * returns 0 on succcess
+ */
+int vfio_get_region_cap_type(struct vfio_region_info *region_info,
+			     mdev_region_class_t *class_info)
+{
+	struct vfio_info_cap_header *caps = NULL;
+	int ret = 0;
+	struct vfio_region_info_cap_type *cap_type;
+
+	if (region_info->flags & VFIO_REGION_INFO_FLAG_CAPS &&
+	    region_info->argsz > sizeof(*region_info)) {
+		caps = vfio_get_cap_info(region_info,
+					 VFIO_REGION_INFO_CAP_TYPE);
+		if (!caps) {
+			ret = -EINVAL;
+			goto out;
+		}
+
+		cap_type =
+		    odp_container_of(caps, struct vfio_region_info_cap_type,
+				     header);
+
+		class_info->type = cap_type->type;
+		class_info->subtype = cap_type->subtype;
+	}
+out:
+	return ret;
+}
+
+/**
+ * Get specific region info
+ */
+static struct vfio_region_info *vfio_get_region(mdev_device_t *mdev,
+						__u32 region)
+{
+	int ret;
+	struct vfio_region_info *region_info = NULL;
+
+	ODP_DBG("Region:%d\n", region);
+	region_info = calloc(1, sizeof(*region_info));
+	if (!region_info)
+		goto out;
+
+	region_info->index = region;
+	region_info->argsz = sizeof(*region_info);
+	ret = ioctl(mdev->device, VFIO_DEVICE_GET_REGION_INFO, region_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get PCI region info\n");
+		goto out;
+	}
+
+	if (!region_info->size) {
+		ODP_DBG("region info %d is empty, skipping\n", region);
```

Comment:
It's not clear why the region not existing results in an `ODP_ERR()` but it being empty is an `ODP_DBG()`. It seems both should be `ODP_DBG()` calls here, since the caller gets a `NULL` return in either case and presumably continues.

Reply:
This call tries to get the area info that's exported from the linux kernel. We loop through all PCI areas which is PCI BARS + Driver defined areas for Rx/Tx descriptors.
Some(usually all but one) PCI BARS are empty, but it's normal. That's why we only issue a debug message here

Reply:
The difference here is because 0-sized region is OK and we don't care about it, but on the other hand we would like to spot GET_REGION_INFO failures even with debug disabled. I would prefer to keep it as is.